File: backend/internal/handler/tasks.go
package handler

import (
	"context"
	"fmt"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/hujinrun/flowspace/internal/model"
	"github.com/hujinrun/flowspace/internal/service"
	"github.com/hujinrun/flowspace/internal/storage"
)

func GetTasks(store storage.Store) gin.HandlerFunc {
	return func(c *gin.Context) {
		page, pageSize := getPagination(c)
		project := c.Query("project")
		status := c.DefaultQuery("status", "all")
		scope := c.Query("scope")
		horizon := c.Query("horizon")
		projectID := c.Query("project_id")
		plannedDate := c.Query("planned_date")
		plannedFrom := c.Query("planned_from")
		plannedTo := c.Query("planned_to")
		executionType := c.Query("execution_type")

		tasks, total, err := service.GetTasks(c.Request.Context(), store, project, status, scope, horizon, projectID, plannedDate, plannedFrom, plannedTo, executionType, page, pageSize)
		if err != nil {
			internalError(c, "failed to get tasks")
			return
		}
		successWithPagination(c, gin.H{"tasks": tasks}, page, pageSize, total)
	}
}

func GetTaskProjects(store storage.Store) gin.HandlerFunc {
	return func(c *gin.Context) {
		projects, err := service.GetTaskProjects(c.Request.Context(), store)
		if err != nil {
			internalError(c, "failed to get task projects")
			return
		}
		success(c, gin.H{"projects": projects})
	}
}

func ListTaskProjects(store storage.Store) gin.HandlerFunc {
	return func(c *gin.Context) {
		projects, err := service.ListTaskProjects(c.Request.Context(), store)
		if err != nil {
			internalError(c, "failed to list task projects")
			return
		}
		success(c, gin.H{"projects": projects})
	}
}

func CreateTaskProject(store storage.Store) gin.HandlerFunc {
	return func(c *gin.Context) {
		var req model.CreateTaskProjectRequest
		if err := c.ShouldBindJSON(&req); err != nil {
			badRequest(c, "project name is required")
			return
		}
		project, err := service.CreateTaskProject(c.Request.Context(), store, &req)
		if err != nil {
			badRequest(c, err.Error())
			return
		}
		created(c, gin.H{"project": project})
	}
}

func UpdateTaskProject(store storage.Store) gin.HandlerFunc {
	return func(c *gin.Context) {
		var req model.UpdateTaskProjectRequest
		if err := c.ShouldBindJSON(&req); err != nil {
			badRequest(c, "invalid request body")
			return
		}
		project, err := service.UpdateTaskProject(c.Request.Context(), store, c.Param("id"), &req)
		if err != nil {
			notFound(c, "project not found")
			return
		}
		success(c, gin.H{"project": project})
	}
}

func DeleteTaskProject(store storage.Store) gin.HandlerFunc {
	return func(c *gin.Context) {
		if err := service.DeleteTaskProject(c.Request.Context(), store, c.Param("id")); err != nil {
			badRequest(c, err.Error())
			return
		}
		noContent(c)
	}
}

func CreateTask(store storage.Store) gin.HandlerFunc {
	return func(c *gin.Context) {
		var req model.CreateTaskRequest
		if err := c.ShouldBindJSON(&req); err != nil {
			badRequest(c, "title is required")
			return
		}
		task, err := service.CreateTask(c.Request.Context(), store, &req)
		if err != nil {
			internalError(c, err.Error())
			return
		}
		created(c, gin.H{"task": task})
	}
}

func UpdateTask(store storage.Store) gin.HandlerFunc {
	return func(c *gin.Context) {
		var req model.UpdateTaskRequest
		if err := c.ShouldBindJSON(&req); err != nil {
			badRequest(c, "invalid request body")
			return
		}
		task, err := service.UpdateTask(c.Request.Context(), store, c.Param("id"), &req)
		if err != nil {
			if rte, ok := err.(*service.RecurringTaskError); ok {
				conflict(c, rte.Code, rte.Message)
				return
			}
			notFound(c, "task not found")
			return
		}
		success(c, gin.H{"task": task})
	}
}

func DeleteTask(store storage.Store) gin.HandlerFunc {
	return func(c *gin.Context) {
		if err := service.DeleteTask(c.Request.Context(), store, c.Param("id")); err != nil {
			internalError(c, "failed to delete task")
			return
		}
		noContent(c)
	}
}

func CompleteOccurrence(store storage.Store) gin.HandlerFunc {
	return func(c *gin.Context) {
		taskID := c.Param("id")
		date := c.Param("date")
		ctx := c.Request.Context()

		if err := validateOccurrenceAction(ctx, store, taskID, date); err != nil {
			badRequest(c, err.Error())
			return
		}

		occurrence, err := store.Recurrence().CompleteOccurrence(ctx, taskID, date, time.Now().Unix())
		if err != nil {
			internalError(c, err.Error())
			return
		}
		success(c, gin.H{"occurrence": occurrence})
	}
}

func ReopenOccurrence(store storage.Store) gin.HandlerFunc {
	return func(c *gin.Context) {
		taskID := c.Param("id")
		date := c.Param("date")
		ctx := c.Request.Context()

		if err := validateOccurrenceAction(ctx, store, taskID, date); err != nil {
			badRequest(c, err.Error())
			return
		}

		occurrence, err := store.Recurrence().ReopenOccurrence(ctx, taskID, date)
		if err != nil {
			internalError(c, err.Error())
			return
		}
		success(c, gin.H{"occurrence": occurrence})
	}
}

func SkipOccurrence(store storage.Store) gin.HandlerFunc {
	return func(c *gin.Context) {
		taskID := c.Param("id")
		date := c.Param("date")
		ctx := c.Request.Context()

		if err := validateOccurrenceAction(ctx, store, taskID, date); err != nil {
			badRequest(c, err.Error())
			return
		}

		occurrence, err := store.Recurrence().SkipOccurrence(ctx, taskID, date)
		if err != nil {
			internalError(c, err.Error())
			return
		}
		success(c, gin.H{"occurrence": occurrence})
	}
}

func GetTaskOccurrences(store storage.Store) gin.HandlerFunc {
	return func(c *gin.Context) {
		from := c.DefaultQuery("from", "")
		to := c.DefaultQuery("to", "")
		if from == "" || to == "" {
			badRequest(c, "from and to query parameters are required")
			return
		}

		occurrences, err := store.Recurrence().ListOccurrences(c.Request.Context(), from, to)
		if err != nil {
			internalError(c, err.Error())
			return
		}
		success(c, gin.H{"occurrences": occurrences})
	}
}

func validateOccurrenceAction(ctx context.Context, store storage.Store, taskID, date string) error {
	task, err := store.Tasks().GetByID(ctx, taskID)
	if err != nil {
		return fmt.Errorf("task not found")
	}
	if task.ExecutionType != "recurring" {
		return fmt.Errorf("task is not a recurring task")
	}

	rule, err := store.Recurrence().GetRule(ctx, taskID)
	if err != nil {
		return fmt.Errorf("recurrence rule not found")
	}

	occurrences := service.ExpandRuleOccurrences(rule, date, date)
	if len(occurrences) == 0 {
		return fmt.Errorf("date is not an expected occurrence")
	}

	return nil
}

File: frontend/src/api/tasks.ts
import { APIError, api } from './client'

export interface Task {
  id: string
  title: string
  content: string
  project?: string
  project_id?: string
  project_type?: string
  due?: number
  planned_date?: string
  priority: number
  done: number
  status: string
  horizon: 'week' | 'long'
  scope: string
  sort_order: number
  note_id?: string
  roadmap_node_id?: string
  execution_type?: 'single' | 'recurring'
  occurrence_date?: string
  occurrence_status?: 'open' | 'done' | 'skipped'
  recurrence_label?: string
  completed_at?: number
  created_at: number
  updated_at: number
}

export interface RecurrenceConfig {
  start_date: string
  end_date?: string
  frequency: 'daily' | 'weekly' | 'monthly'
  interval?: number
  weekdays?: number[]
  month_days?: number[]
  timezone?: string
  enabled?: boolean
}

export interface TaskOccurrence {
  task_id: string
  occurrence_date: string
  status: 'open' | 'done' | 'skipped'
  completed_at?: number
  title: string
  content: string
  project_id?: string
  project?: string
  roadmap_node_id?: string
  recurrence_label: string
}

export interface TaskProject {
  id: string
  name: string
  type: 'personal' | 'regular' | 'learning'
  description: string
  created_at: number
  updated_at: number
}

export interface LearningRoadmap {
  id: string
  project_id: string
  title: string
  goal: string
  status: 'draft' | 'ready' | 'failed'
  nodes: RoadmapNode[]
  edges: RoadmapEdge[]
  created_at: number
  updated_at: number
}

export interface RoadmapNode {
  id: string
  roadmap_id: string
  parent_id?: string
  type: 'phase' | 'module' | 'task' | 'choice'
  title: string
  description: string
  path_type: 'required' | 'recommended' | 'optional' | 'alternative'
  status: 'todo' | 'active' | 'done' | 'skipped'
  deliverable: string
  acceptance_criteria: string
  x: number
  y: number
  order_index: number
  article_search_queries?: string[]
  resources: RoadmapResource[]
  created_at: number
  updated_at: number
}

export interface RoadmapEdge {
  id: string
  roadmap_id: string
  source_node_id: string
  target_node_id: string
  style: 'solid' | 'dotted'
  created_at: number
}

export interface RoadmapResource {
  id: string
  node_id: string
  title: string
  url: string
  summary: string
  source_type: string
  added_by: string
  created_at: number
  updated_at: number
}

export async function getTasks(params: {
  project?: string
  project_id?: string
  status?: string
  scope?: string
  horizon?: string
  planned_date?: string
  page?: number
  page_size?: number
}) {
  const res = await api.get<{ tasks: Task[] }>('/api/tasks', {
    project: params.project ?? '',
    project_id: params.project_id ?? '',
    status: params.status ?? 'all',
    scope: params.scope ?? '',
    horizon: params.horizon ?? '',
    planned_date: params.planned_date ?? '',
    page: String(params.page ?? 1),
    page_size: String(params.page_size ?? 100),
  })
  return { tasks: res.data.tasks, pagination: res.pagination! }
}

export async function getTaskProjects() {
  try {
    const res = await api.get<{ projects: string[] }>('/api/tasks/projects')
    return res.data.projects
  } catch (error) {
    if (error instanceof APIError && error.status === 404) return []
    throw error
  }
}

export async function listTaskProjects() {
  const res = await api.get<{ projects: TaskProject[] }>('/api/task-projects')
  return res.data.projects
}

export async function createTaskProject(body: {
  name: string
  type: TaskProject['type']
  description?: string
}) {
  const res = await api.post<{ project: TaskProject }>(
    '/api/task-projects',
    body
  )
  return res.data.project
}

export async function updateTaskProject(
  id: string,
  body: Partial<Pick<TaskProject, 'name' | 'type' | 'description'>>
) {
  const res = await api.patch<{ project: TaskProject }>(
    `/api/task-projects/${id}`,
    body
  )
  return res.data.project
}

export async function deleteTaskProject(id: string) {
  await api.del(`/api/task-projects/${id}`)
}

export async function createTask(body: {
  title: string
  content?: string
  project?: string
  project_id?: string
  due?: number
  planned_date?: string
  priority?: number
  scope?: string
  horizon?: 'week' | 'long'
  roadmap_node_id?: string
  execution_type?: 'single' | 'recurring'
  recurrence?: RecurrenceConfig
}) {
  const res = await api.post<{ task: Task }>('/api/tasks', body)
  return res.data.task
}

export async function updateTask(id: string, body: Partial<Task>) {
  const res = await api.patch<{ task: Task }>(`/api/tasks/${id}`, body)
  return res.data.task
}

export async function deleteTask(id: string) {
  await api.del(`/api/tasks/${id}`)
}

export async function generateLearningRoadmap(projectID: string) {
  const res = await api.post<{ roadmap: LearningRoadmap }>(
    `/api/task-projects/${projectID}/roadmap/generate`
  )
  return res.data.roadmap
}

export async function getLearningRoadmap(projectID: string) {
  try {
    const res = await api.get<{ roadmap: LearningRoadmap }>(
      `/api/task-projects/${projectID}/roadmap`
    )
    return res.data.roadmap
  } catch (error) {
    if (error instanceof APIError && error.status === 404) return null
    throw error
  }
}

export async function updateRoadmapNode(
  id: string,
  body: Partial<RoadmapNode>
) {
  const res = await api.patch<{ node: RoadmapNode }>(
    `/api/roadmap-nodes/${id}`,
    body
  )
  return res.data.node
}

export async function createRoadmapNode(
  roadmapID: string,
  body: {
    parent_id?: string
    title: string
    type?: RoadmapNode['type']
    description?: string
    path_type?: RoadmapNode['path_type']
    status?: RoadmapNode['status']
    deliverable?: string
    acceptance_criteria?: string
    edge_style?: 'solid' | 'dotted'
  }
) {
  const res = await api.post<{ node: RoadmapNode }>(
    `/api/roadmaps/${roadmapID}/nodes`,
    body
  )
  return res.data.node
}

export async function deleteRoadmapNode(id: string) {
  await api.del(`/api/roadmap-nodes/${id}`)
}

export async function saveRoadmapLayout(
  roadmapID: string,
  nodes: Array<{ id: string; x: number; y: number }>
) {
  await api.patch(`/api/roadmaps/${roadmapID}/layout`, { nodes })
}

export async function optimizeRoadmapLayout(roadmapID: string) {
  const res = await api.post<{ roadmap: LearningRoadmap }>(
    `/api/roadmaps/${roadmapID}/layout/optimize`
  )
  return res.data.roadmap
}

export async function searchRoadmapNodeResources(
  nodeID: string,
  body: { sources?: string[] } = {}
) {
  const res = await api.post<{
    node_id: string
    query: string
    resources: RoadmapResource[]
  }>(`/api/roadmap-nodes/${nodeID}/resources/search`, body)
  return res.data
}

export async function addRoadmapNodeResource(
  nodeID: string,
  body: { title: string; url: string; summary?: string }
) {
  const res = await api.post<{ resource: RoadmapResource }>(
    `/api/roadmap-nodes/${nodeID}/resources`,
    body
  )
  return res.data.resource
}

export async function deleteRoadmapResource(id: string) {
  await api.del(`/api/roadmap-resources/${id}`)
}

export async function completeOccurrence(taskId: string, date: string) {
  const res = await api.post<{ occurrence: TaskOccurrence }>(
    `/api/tasks/${taskId}/occurrences/${date}/complete`
  )
  return res.data.occurrence
}

export async function reopenOccurrence(taskId: string, date: string) {
  const res = await api.post<{ occurrence: TaskOccurrence }>(
    `/api/tasks/${taskId}/occurrences/${date}/reopen`
  )
  return res.data.occurrence
}

export async function skipOccurrence(taskId: string, date: string) {
  const res = await api.post<{ occurrence: TaskOccurrence }>(
    `/api/tasks/${taskId}/occurrences/${date}/skip`
  )
  return res.data.occurrence
}

export async function getTaskOccurrences(from: string, to: string) {
  const res = await api.get<{ occurrences: TaskOccurrence[] }>(
    '/api/task-occurrences',
    { from, to }
  )
  return res.data.occurrences
}

export async function getRecurringTasks(params?: {
  page?: number
  page_size?: number
}) {
  const res = await api.get<{ tasks: Task[] }>('/api/tasks', {
    execution_type: 'recurring',
    page: String(params?.page ?? 1),
    page_size: String(params?.page_size ?? 100),
  })
  return { tasks: res.data.tasks, pagination: res.pagination! }
}
